### PR TITLE
add type annotations

### PR DESCRIPTION
**Pull request by joostvanzwieten**

### CODE REVIEW

File: treelog/_io.py [hunk 1]
```diff
@@ -18,53 +18,37 @@
...
-  def open(self, filename, mode, *, encoding=None, name=None, umask=0o666):
-    if mode == 'w' or mode == 'w+':
-      wrapper = functools.partial(io.TextIOWrapper, encoding=encoding)
-    elif mode == 'wb':
-      wrapper = io.BufferedWriter
-    elif mode == 'wb+':
-      wrapper = io.BufferedRandom
-    else:
+  def open(self, filename: str, mode: str, *, encoding: typing.Optional[str] = None, umask: int = 0o666) -> proto.IO:
+    if mode not in ('w', 'w+', 'wb', 'wb+'):
```
Comment (gertjanvanzwieten) [inline treelog/_io.py]: let's limit this to w, wb (similar to module)

File: tests.py [hunk 2]
```diff
@@ -60,9 +61,9 @@ class StdoutLog(Log):
 
   @contextlib.contextmanager
   def output_tester(self):
-    with capture() as writes:
+    with capture() as captured:
       yield treelog.StdoutLog()
-    self.assertEqual(''.join(writes),
+    self.assertEqual(''.join(captured.stdout),
```
Comment (gertjanvanzwieten) [inline tests.py]: unnecessary join

File: treelog/iter.py [hunk 3]
```diff
@@ -1,58 +1,61 @@
...
   next item a new log context is opened taken from the ``titles`` iterable. The
   wrapped object should be entered before use in order to ensure that this
   context is properly closed in case the iterator is prematurely abandoned.'''
 
-  def __init__(self, titles, iterable):
+  def __init__(self, titles: typing.Union[typing.Iterable[str], typing.Generator[str, T, None]], iterable: typing.Iterable[T]) -> None:
     self._titles = iter(titles)
     self._iterable = iter(iterable)
-    self._log = None
+    self._log = True # type: typing.Union[proto.Log, bool]
```
Comment (gertjanvanzwieten) [inline treelog/iter.py]: why not just typing.Optional and keep _log at None?

Comment (joostvanzwieten) [reply treelog/iter.py]: In the old code `self._log` could be `None` (not yet entered), a `Log` instance (entered) or `False` (exited). Replacing `None` with `True` reduces the number of types to `proto.Log` and `bool` and arguably simplifies the tests in `__iter__` and `__exit__`.

**Comment by gertjanvanzwieten** [comment]
What if we set it back to None? I don't think there's an issue with
entering it again after it's been closed.

File: treelog/iter.py (hunk 3 above)
Comment (joostvanzwieten) [reply treelog/iter.py]: Done.

**Pull request merged**